broke-broker/src/main/java/org/dsngroup/broke/broker/channel/handler/MessagePublisher.java

```diff
@@ -0,0 +1,90 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dsngroup.broke.broker.channel.handler;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.Channel;
+import org.dsngroup.broke.broker.ServerContext;
+import org.dsngroup.broke.broker.storage.MessagePool;
+import org.dsngroup.broke.broker.storage.ServerSession;
+import org.dsngroup.broke.protocol.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MessagePublisher {
+
+    private static final Logger logger = LoggerFactory.getLogger(MessagePublisher.class);
+
+    public void processQos0Publish(Channel channel, ServerContext ctx, MqttPublishMessage mqttPublishMessage) {
+        // TODO
+    }
+
+    public void processQos1Publish(Channel channel, ServerContext serverContext, MqttPublishMessage mqttPublishMessage) {
+
+        try {
+            if(channel.isActive()) {
+                MessagePool messagePool = serverContext.getMessagePool();
+
+                String topic = mqttPublishMessage.variableHeader().topicName();
+                int packetId = mqttPublishMessage.variableHeader().packetId();
+                ByteBuf appMessage = mqttPublishMessage.payload();
+
+                if (appMessage.isReadable()) {
+                    messagePool.putContentOnTopic(topic, appMessage);
+                } else {
+                    // TODO: handle unreadable payload
+                }
+
+                publishToSubscriptions(serverContext, mqttPublishMessage);
+
+                MqttPubAckMessage mqttPubAckMessage = pubAck(channel, MqttQoS.AT_LEAST_ONCE, packetId);
+                channel.writeAndFlush(mqttPubAckMessage);
+
+            } else {
+                logger.error("Inactive channel");
+            }
+        } catch (Exception e) {
+            logger.error(e.getMessage());
+        }
+
+    }
+
+    private void publishToSubscriptions(ServerContext serverContext,
+                                      MqttPublishMessage mqttPublishMessage) {
+        // TODO: publish messages to the subscriptions in every server session
+        // Iterate through all sessions, publishing to every sessions' subscriptions
+        for (ServerSession serverSession: serverContext.getServerSessionPool().asCollection()) {
+            // Only publish to active sessions
+            if(serverSession.getIsActive()) {
+                // Whether the mqttPublish Message matches the subscription is performed in "sendToSubscribers"
+                serverSession.getSubscriptionPool().sendToSubscribers(mqttPublishMessage);
+            }
```

Comment:
Async

Reply:
I'm considering making this asynchronous (or multi-threaded) will conflict with the EventLoopGroup design of Netty. 

For each channel, an event loop is assigned to run all handlers on such channel, which will be run by a single thread. So the channels' tasks won't block each other unless the threads assigned to the event loop group is less than the number of channels. 

Making this asynchronous means that more than one threads are added to a **single** channel, which may make the context switch more often if there are multiple channels. 

In summary, multiple channel handlers won't be run by different threads since they are in a single channel. Only different channels will be run by different threads.

Reply:
No, actually we don’t need to add additional threads to deal with this. IMHO, the write back to subscriber may need to register an additional handler as outbound in pipeline. The chosen time and scheduling will be assigned by netty. It’s not clear I mark async here, instead, the place of write subscriber’s channel. You should figure out the correct non-blocking read/write in netty.

Reply:
OK, I see!